CPUParticles2D in gles2 gives white screen when particles_animation in material is checked

<!-- Please search existing issues for potential duplicates before filing yours:
https://github.com/godotengine/godot/issues?q=is%3Aissue
-->

**Godot version:**
<!-- Specify commit hash if non-official. -->
869887641f3d31f19837a34da116039a26618cb0

**OS/device including version:**
<!-- Specify GPU model and drivers if graphics-related. -->
Ubuntu 18.04.1 LTS 64bit 
GeForce GT 540M/PCIe/SSE2
driver : 390.77

**Issue description:**
<!-- What happened, and what was expected. -->
Screen turns to white if "particles animation" is check in the material of the CPUParticle2D node is checked.
In gles3 is working as expected
**Steps to reproduce:**
^^^

**Minimal reproduction project:**
<!-- Recommended as it greatly speeds up debugging. Drag and drop a zip archive to upload it. -->
[gles2_bug.zip](https://github.com/godotengine/godot/files/2824205/gles2_bug.zip)

@JFonS Still the same.

Error from terminal : 
<details><pre>
1: #version 120
2: #define USE_GLES_OVER_GL
3: #define USE_INSTANCING
4: #define USE_INSTANCE_CUSTOM
5: 
6: #ifdef USE_GLES_OVER_GL
7: #define lowp
8: #define mediump
9: #define highp
10: #else
11: precision highp float;
12: precision highp int;
13: #endif
14: 
15: uniform highp mat4 projection_matrix;
16: /* clang-format on */
17: 
18: 
19: vec2 select2(vec2 a, vec2 b, bvec2 c) {
20: 	vec2 ret;
21: 
22: 	ret.x = c.x ? b.x : a.x;
23: 	ret.y = c.y ? b.y : a.y;
24: 
25: 	return ret;
26: }
27: 
28: vec3 select3(vec3 a, vec3 b, bvec3 c) {
29: 	vec3 ret;
30: 
31: 	ret.x = c.x ? b.x : a.x;
32: 	ret.y = c.y ? b.y : a.y;
33: 	ret.z = c.z ? b.z : a.z;
34: 
35: 	return ret;
36: }
37: 
38: vec4 select4(vec4 a, vec4 b, bvec4 c) {
39: 	vec4 ret;
40: 
41: 	ret.x = c.x ? b.x : a.x;
42: 	ret.y = c.y ? b.y : a.y;
43: 	ret.z = c.z ? b.z : a.z;
44: 	ret.w = c.w ? b.w : a.w;
45: 
46: 	return ret;
47: }
48: 
49: highp vec4 texel2DFetch(highp sampler2D tex, ivec2 size, ivec2 coord) {
50: 	float x_coord = float(2 * coord.x + 1) / float(size.x * 2);
51: 	float y_coord = float(2 * coord.y + 1) / float(size.y * 2);
52: 
53: 	return texture2DLod(tex, vec2(x_coord, y_coord), 0.0);
54: }
55: 
56: #ifndef USE_GLES_OVER_GL
57: highp mat4 transpose(highp mat4 src) {
58: 	return mat4(
59: 			vec4(src[0].x, src[1].x, src[2].x, src[3].x),
60: 			vec4(src[0].y, src[1].y, src[2].y, src[3].y),
61: 			vec4(src[0].z, src[1].z, src[2].z, src[3].z),
62: 			vec4(src[0].w, src[1].w, src[2].w, src[3].w));
63: }
64: #endif
65: 
66: uniform highp mat4 modelview_matrix;
67: uniform highp mat4 extra_matrix;
68: attribute highp vec2 vertex; // attrib:0
69: attribute vec4 color_attrib; // attrib:3
70: attribute vec2 uv_attrib; // attrib:4
71: 
72: #ifdef USE_SKELETON
73: attribute highp vec4 bone_indices; // attrib:6
74: attribute highp vec4 bone_weights; // attrib:7
75: #endif
76: 
77: #ifdef USE_INSTANCING
78: 
79: attribute highp vec4 instance_xform0; //attrib:8
80: attribute highp vec4 instance_xform1; //attrib:9
81: attribute highp vec4 instance_xform2; //attrib:10
82: attribute highp vec4 instance_color; //attrib:11
83: 
84: #ifdef USE_INSTANCE_CUSTOM
85: attribute highp vec4 instance_custom_data; //attrib:12
86: #endif
87: 
88: #endif
89: 
90: #ifdef USE_SKELETON
91: uniform highp sampler2D skeleton_texture; // texunit:-3
92: uniform highp ivec2 skeleton_texture_size;
93: uniform highp mat4 skeleton_transform;
94: uniform highp mat4 skeleton_transform_inverse;
95: #endif
96: 
97: varying vec2 uv_interp;
98: varying vec4 color_interp;
99: 
100: uniform highp vec2 color_texpixel_size;
101: 
102: #ifdef USE_TEXTURE_RECT
103: 
104: uniform vec4 dst_rect;
105: uniform vec4 src_rect;
106: 
107: #endif
108: 
109: uniform highp float time;
110: 
111: #ifdef USE_LIGHTING
112: 
113: // light matrices
114: uniform highp mat4 light_matrix;
115: uniform highp mat4 light_matrix_inverse;
116: uniform highp mat4 light_local_matrix;
117: uniform highp mat4 shadow_matrix;
118: uniform highp vec4 light_color;
119: uniform highp vec4 light_shadow_color;
120: uniform highp vec2 light_pos;
121: uniform highp float shadowpixel_size;
122: uniform highp float shadow_gradient;
123: uniform highp float light_height;
124: uniform highp float light_outside_alpha;
125: uniform highp float shadow_distance_mult;
126: 
127: varying vec4 light_uv_interp;
128: varying vec2 transformed_light_uv;
129: varying vec4 local_rot;
130: 
131: #ifdef USE_SHADOWS
132: varying highp vec2 pos;
133: #endif
134: 
135: const bool at_light_pass = true;
136: #else
137: const bool at_light_pass = false;
138: #endif
139: 
140: /* clang-format off */
141: uniform int m_particles_anim_h_frames;
142: uniform int m_particles_anim_v_frames;
143: uniform bool m_particles_anim_loop;
144: 
145: 
146: /* clang-format on */
147: 
148: vec2 select(vec2 a, vec2 b, bvec2 c) {
149: 	vec2 ret;
150: 
151: 	ret.x = c.x ? b.x : a.x;
152: 	ret.y = c.y ? b.y : a.y;
153: 
154: 	return ret;
155: }
156: 
157: void main() {
158: 
159: 	vec4 color = color_attrib;
160: 
161: #ifdef USE_INSTANCING
162: 	mat4 extra_matrix_instance = extra_matrix * transpose(mat4(instance_xform0, instance_xform1, instance_xform2, vec4(0.0, 0.0, 0.0, 1.0)));
163: 	color *= instance_color;
164: 	vec4 instance_custom = instance_custom_data;
165: 
166: #else
167: 	mat4 extra_matrix_instance = extra_matrix;
168: 	vec4 instance_custom = vec4(0.0);
169: #endif
170: 
171: #ifdef USE_TEXTURE_RECT
172: 
173: 	if (dst_rect.z < 0.0) { // Transpose is encoded as negative dst_rect.z
174: 		uv_interp = src_rect.xy + abs(src_rect.zw) * vertex.yx;
175: 	} else {
176: 		uv_interp = src_rect.xy + abs(src_rect.zw) * vertex;
177: 	}
178: 
179: 	vec4 outvec = vec4(0.0, 0.0, 0.0, 1.0);
180: 
181: 	// This is what is done in the GLES 3 bindings and should
182: 	// take care of flipped rects.
183: 	//
184: 	// But it doesn't.
185: 	// I don't know why, will need to investigate further.
186: 
187: 	outvec.xy = dst_rect.xy + abs(dst_rect.zw) * select(vertex, vec2(1.0, 1.0) - vertex, lessThan(src_rect.zw, vec2(0.0, 0.0)));
188: 
189: 	// outvec.xy = dst_rect.xy + abs(dst_rect.zw) * vertex;
190: #else
191: 	vec4 outvec = vec4(vertex.xy, 0.0, 1.0);
192: 
193: 	uv_interp = uv_attrib;
194: #endif
195: 
196: 	{
197: 		vec2 src_vtx = outvec.xy;
198: 		/* clang-format off */
199: {
200: 	float m_h_frames = float(m_particles_anim_h_frames);
201: 	float m_v_frames = float(m_particles_anim_v_frames);
202: 	outvec.xy.xy /= vec2(m_h_frames, m_v_frames);
203: 	int m_total_frames = (m_particles_anim_h_frames * m_particles_anim_v_frames);
204: 	int m_frame = int((float(m_total_frames) * instance_custom.z));
205: 	if (m_particles_anim_loop)
206: 	{
207: 			{
208: 		m_frame = int(mod(float(abs(m_frame)), float(m_total_frames)));
209: 	}
210: ;
211: 	}
212: 	else
213: 	{
214: 			{
215: 		m_frame = clamp(m_frame, 0, (m_total_frames - 1));
216: 	}
217: ;
218: 	}
219: 	float m_frame_w = (1.0 / m_h_frames);
220: 	float m_frame_h = (1.0 / m_v_frames);
221: 	uv_interp.x = ((uv_interp.x * m_frame_w) + (m_frame_w * float(mod(float(m_frame), float(m_particles_anim_h_frames)))));
222: 	uv_interp.y = ((uv_interp.y * m_frame_h) + (m_frame_h * float((m_frame / m_particles_anim_h_frames))));
223: }
224: 
225: 
226: 		/* clang-format on */
227: 	}
228: 
229: #if !defined(SKIP_TRANSFORM_USED)
230: 	outvec = extra_matrix_instance * outvec;
231: 	outvec = modelview_matrix * outvec;
232: #endif
233: 
234: 	color_interp = color;
235: 
236: #ifdef USE_PIXEL_SNAP
237: 	outvec.xy = floor(outvec + 0.5).xy;
238: #endif
239: 
240: #ifdef USE_SKELETON
241: 
242: 	// look up transform from the "pose texture"
243: 	if (bone_weights != vec4(0.0)) {
244: 
245: 		highp mat4 bone_transform = mat4(0.0);
246: 
247: 		for (int i = 0; i < 4; i++) {
248: 			ivec2 tex_ofs = ivec2(int(bone_indices[i]) * 2, 0);
249: 
250: 			highp mat4 b = mat4(
251: 					texel2DFetch(skeleton_texture, skeleton_texture_size, tex_ofs + ivec2(0, 0)),
252: 					texel2DFetch(skeleton_texture, skeleton_texture_size, tex_ofs + ivec2(1, 0)),
253: 					vec4(0.0, 0.0, 1.0, 0.0),
254: 					vec4(0.0, 0.0, 0.0, 1.0));
255: 
256: 			bone_transform += b * bone_weights[i];
257: 		}
258: 
259: 		mat4 bone_matrix = skeleton_transform * transpose(bone_transform) * skeleton_transform_inverse;
260: 
261: 		outvec = bone_matrix * outvec;
262: 	}
263: 
264: #endif
265: 
266: 	gl_Position = projection_matrix * outvec;
267: 
268: #ifdef USE_LIGHTING
269: 
270: 	light_uv_interp.xy = (light_matrix * outvec).xy;
271: 	light_uv_interp.zw = (light_local_matrix * outvec).xy;
272: 
273: 	transformed_light_uv = (mat3(light_matrix_inverse) * vec3(light_uv_interp.zw, 0.0)).xy; //for normal mapping
274: 
275: #ifdef USE_SHADOWS
276: 	pos = outvec.xy;
277: #endif
278: 
279: 	local_rot.xy = normalize((modelview_matrix * (extra_matrix_instance * vec4(1.0, 0.0, 0.0, 0.0))).xy);
280: 	local_rot.zw = normalize((modelview_matrix * (extra_matrix_instance * vec4(0.0, 1.0, 0.0, 0.0))).xy);
281: #ifdef USE_TEXTURE_RECT
282: 	local_rot.xy *= sign(src_rect.z);
283: 	local_rot.zw *= sign(src_rect.w);
284: #endif
285: 
286: #endif
287: }
288: 
289: /* clang-format off */
290: 
ERROR: _display_error_with_code: CanvasShaderGLES2: Vertex shader compilation failed:
0(208) : error C1101: ambiguous overloaded function reference "abs(int)"
    (0) : gp5 cp50 fp50 vp50 gp50 cpf fpf vpf gpf superp float abs(superp float)
    (0) : ps gp5 gp4 cp50 fp50 vp50 gp50 cpf fpf vpf gpf lowp float abs(lowp float)
    (0) : ps gp5 gp4 cp50 fp50 vp50 gp50 cpf fpf vpf gpf mediump float abs(mediump float)
    (0) : float abs(float)
0(215) : error C7011: implicit cast from "float" to "int"
</pre></deatails>

--- a/scene/2d/canvas_item.cpp
+++ b/scene/2d/canvas_item.cpp
@@ -131,19 +131,15 @@ void CanvasItemMaterial::_update_shader() {
 
 		code += "\tVERTEX.xy /= vec2(h_frames, v_frames);\n";
 
-		code += "\tint total_frames = particles_anim_h_frames * particles_anim_v_frames;\n";
-		code += "\tint frame = int(float(total_frames) * INSTANCE_CUSTOM.z);\n";
-		code += "\tif (particles_anim_loop) {\n";
-		code += "\t\tframe = int(abs(frame) % total_frames);\n";
+		code += "\tfloat particle_total_frames = float(particles_anim_h_frames * particles_anim_v_frames);\n";
+		code += "\tfloat particle_frame = floor(INSTANCE_CUSTOM.z * float(particle_total_frames));\n";
+		code += "\tif (!particles_anim_loop) {\n";
+		code += "\t\tparticle_frame = clamp(particle_frame, 0.0, particle_total_frames - 1.0);\n";
 		code += "\t} else {\n";
-		code += "\t\tframe = clamp(frame, 0, total_frames - 1);\n";
-		code += "\t}\n";
-
-		code += "\tfloat frame_w = 1.0 / h_frames;\n";
-		code += "\tfloat frame_h = 1.0 / v_frames;\n";
-		code += "\tUV.x = UV.x * frame_w + frame_w * float(frame % particles_anim_h_frames);\n";
-		code += "\tUV.y = UV.y * frame_h + frame_h * float(frame / particles_anim_h_frames);\n";
-
+		code += "\t\tparticle_frame = mod(particle_frame, particle_total_frames);\n";
+		code += "\t}";
+		code += "\tUV /= vec2(h_frames, v_frames);\n";
+		code += "\tUV += vec2(mod(particle_frame, h_frames) / h_frames, floor(particle_frame / h_frames) / v_frames);\n";
 		code += "}\n";
 	}
 

--- a/scene/resources/material.cpp
+++ b/scene/resources/material.cpp
@@ -617,11 +617,17 @@ void SpatialMaterial::_update_shader() {
 			code += "\tMODELVIEW_MATRIX = INV_CAMERA_MATRIX * mat_world;\n";
 
 			//handle animation
+			code += "\tfloat h_frames = float(particles_anim_h_frames);\n";
+			code += "\tfloat v_frames = float(particles_anim_v_frames);\n";
 			code += "\tfloat particle_total_frames = float(particles_anim_h_frames * particles_anim_v_frames);\n";
 			code += "\tfloat particle_frame = floor(INSTANCE_CUSTOM.z * float(particle_total_frames));\n";
-			code += "\tif (!particles_anim_loop) particle_frame=clamp(particle_frame,0.0,particle_total_frames-1.0); else particle_frame=mod(particle_frame,float(particle_total_frames));\n";
-			code += "\tUV /= vec2(float(particles_anim_h_frames),float(particles_anim_v_frames));\n";
-			code += "\tUV += vec2(mod(particle_frame,float(particles_anim_h_frames)) / float(particles_anim_h_frames), floor(particle_frame / float(particles_anim_h_frames)) / float(particles_anim_v_frames));\n";
+			code += "\tif (!particles_anim_loop) {\n";
+			code += "\t\tparticle_frame = clamp(particle_frame, 0.0, particle_total_frames - 1.0);\n";
+			code += "\t} else {\n";
+			code += "\t\tparticle_frame = mod(particle_frame, particle_total_frames);\n";
+			code += "\t}";
+			code += "\tUV /= vec2(h_frames, v_frames);\n";
+			code += "\tUV += vec2(mod(particle_frame, h_frames) / h_frames, floor(particle_frame / h_frames) / v_frames);\n";
 		} break;
 	}
 

